samples/nrf_desktop/src/modules/hid_forward.c

```diff
@@ -0,0 +1,221 @@
+static const void *usb_id;
+static atomic_t usb_ready;
+static struct hid_mouse_event *next_event;
+static bool usb_busy;
+
+
+static u8_t hidc_read(struct bt_gatt_hids_c *hids_c,
+		      struct bt_gatt_hids_c_rep_info *rep,
+		      u8_t err,
+		      const u8_t *data)
+{
+	if (!data) {
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (err || !atomic_get(&usb_ready)) {
+		return BT_GATT_ITER_CONTINUE;
+	}
+
+	u8_t button_bm = data[0];
+	s16_t wheel = (s8_t)data[1];
+
+	u8_t x_buff[2] = {data[2], data[3] & 0x0F};
+	u8_t y_buff[2] = {(data[3] >> 4) | ((data[4] << 4) & 0xF0), data[4] >> 4};
+
+	u16_t x = sys_get_le16(x_buff);
+	u16_t y = sys_get_le16(y_buff);
+
+	if (x > REPORT_MOUSE_XY_MAX) {
+		x |= 0xF000;
+	}
+	if (y > REPORT_MOUSE_XY_MAX) {
+		y |= 0xF000;
+	}
+
+	struct hid_mouse_event *event;
+
+	u32_t key = irq_lock();
+
+	if (next_event) {
+		__ASSERT_NO_MSG(usb_busy);
+
+		LOG_WRN("Event override");
+		event = next_event;
+	} else {
+		event = new_hid_mouse_event();
+	}
+
+	event->subscriber = usb_id;
+
+	event->button_bm = button_bm;
+	event->wheel     = wheel;
+
+	event->dx = x;
+	event->dy = y;
+
+	if (!usb_busy) {
+		EVENT_SUBMIT(event);
+		usb_busy = true;
+	} else {
+		next_event = event;
+	}
+
+	irq_unlock(key);
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+static void hidc_ready(struct bt_gatt_hids_c *hids_c)
+{
+	int err = bt_gatt_hids_c_rep_subscribe(&hidc, hidc.rep_info[0],
+					       hidc_read);
+
+	if (err) {
+		LOG_ERR("Cannot subscribe to report (err:%d)", err);
+	}
+}
+
+static void hidc_pm_update(struct bt_gatt_hids_c *hids_c)
+{
+	LOG_INF("Protocol mode updated");
+}
+
+static void hidc_prep_error(struct bt_gatt_hids_c *hids_c, int err)
+{
+	if (err) {
+		LOG_ERR("err:%d", err);
+	}
+}
+
+static void init(void)
+{
+	static const struct bt_gatt_hids_c_init_params params = {
+		.ready_cb = hidc_ready,
+		.prep_error_cb = hidc_prep_error,
+		.pm_update_cb = hidc_pm_update,
+	};
+
+	bt_gatt_hids_c_init(&hidc, &params);
+}
+
+static int assign_handles(struct bt_gatt_dm *dm)
+{
+	int err = bt_gatt_hids_c_handles_assign(dm, &hidc);
+
+	if (err) {
+		LOG_ERR("Cannot assign handles (err:%d)", err);
+	}
+
+	return err;
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (is_hid_report_sent_event(eh)) {
+		u32_t key = irq_lock();
+		if (next_event) {
+			EVENT_SUBMIT(next_event);
+			next_event = NULL;
+		} else {
+			usb_busy = false;
+		}
+		irq_unlock(key);
+
+		return false;
+	}
+
+	if (is_module_state_event(eh)) {
+		const struct module_state_event *event =
+			cast_module_state_event(eh);
+
+		if (check_state(event, MODULE_ID(ble_state),
+				MODULE_STATE_READY)) {
+			static bool initialized;
+
+			__ASSERT_NO_MSG(!initialized);
+			initialized = true;
+
+			init();
+			module_set_state(MODULE_STATE_READY);
+		}
+
+		return false;
+	}
+
+	if (is_ble_discovery_complete_event(eh)) {
+		const struct ble_discovery_complete_event *event =
+			cast_ble_discovery_complete_event(eh);
+
+		LOG_ERR("");
```

Comment:
Empty log message, why getting this event is an error?

Reply:
debug leftover, thanks